core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java

```diff
@@ -109,50 +112,58 @@ private void setPrimariesInitialRecoveries(int primariesInitialRecoveries) {
+                return allocation.decision(THROTTLE, NAME, "too many incoming shards are currently recovering [%d], limit: [%d]",
+                    currentInRecoveries, concurrentIncomingRecoveries);
+            } else {
+                // search for corresponding recovery source (= primary shard) and check number of outgoing recoveries on that node
+                ShardRouting primaryShard = allocation.routingNodes().activePrimary(shardRouting.shardId());
+                if (primaryShard == null) {
+                    return allocation.decision(Decision.NO, NAME, "primary shard for this replica is not yet active");
                 }
-                if (primariesInRecovery >= primariesInitialRecoveries) {
-                    return allocation.decision(Decision.THROTTLE, NAME, "too many primaries are currently recovering [%d], limit: [%d]",
-                            primariesInRecovery, primariesInitialRecoveries);
+                int primaryNodeOutRecoveries = allocation.routingNodes().getOutgoingRecoveries(primaryShard.currentNodeId());
+                if (primaryNodeOutRecoveries >= concurrentOutgoingRecoveries) {
+                    return allocation.decision(THROTTLE, NAME, "too many outgoing shards are currently recovering [%d], limit: [%d]",
+                        primaryNodeOutRecoveries, concurrentOutgoingRecoveries);
                 } else {
-                    return allocation.decision(Decision.YES, NAME, "below primary recovery limit of [%d]", primariesInitialRecoveries);
+                    return allocation.decision(YES, NAME, "below shard recovery limit of outgoing: [%d < %d] incoming: [%d < %d]",
+                        primaryNodeOutRecoveries,
+                        concurrentOutgoingRecoveries,
+                        currentInRecoveries,
+                        concurrentIncomingRecoveries);
                 }
             }
-        }
-        // TODO should we allow shards not allocated post API to always allocate?
-        // either primary or replica doing recovery (from peer shard)
+        } else {
+            // primary is unassigned, means we are going to do recovery from store, snapshot or local shards
+            // count *just the primaries* currently doing recovery on the node and check against primariesInitialRecoveries
+            assert shardRouting.primary() && shardRouting.unassigned();
 
-        // count the number of recoveries on the node, its for both target (INITIALIZING) and source (RELOCATING)
-        return canAllocate(node, allocation);
-    }
+            // TODO: Should index creation not be throttled?
+//            if (shardRouting.unassignedInfo().getReason() == UnassignedInfo.Reason.INDEX_CREATED) {
+//                return allocation.decision(Decision.YES, NAME, "empty shard does not need recovery");
+//            }
 
-    @Override
-    public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
```

Comment:
Yeah, I wasn't too happy on removing it. Unfortunately, it wasn't safe and there is no safe version which makes sense. Even if `incomingRecoveries` is over the limit, we might still allocate primary unassigned shards (which are only bound by `primariesInitialRecoveries`). Checking for `outgoingRecoveries` on this node to allocate a shard makes no sense at all, as allocating a shard to this node only affects the outgoing recoveries of the node that has the primary shard.